File: 17-Chain-of-Responsibility/src/go/approver.go
package main

import "fmt"

// Approver is the Handler interface. Each handler can process a request
// or forward it, and can be linked to a successor.
type Approver interface {
	Handle(request PurchaseRequest)
	SetNext(next Approver) Approver
}

// BaseApprover supplies the shared next-reference and forwarding helper.
// Concrete handlers embed it and implement their own Handle.
type BaseApprover struct {
	next Approver
}

func (b *BaseApprover) SetNext(next Approver) Approver {
	b.next = next
	return next
}

// passToNext forwards the request, or reports that no one could approve it.
func (b *BaseApprover) passToNext(request PurchaseRequest) {
	if b.next != nil {
		b.next.Handle(request)
	} else {
		fmt.Printf("  Request for $%.2f (%s) could not be approved by anyone.\n",
			request.Amount, request.Purpose)
	}
}

File: 17-Chain-of-Responsibility/src/rust/approver.rs
use crate::purchase_request::PurchaseRequest;

/// Handler trait. Concrete handlers supply their limit, name, and successor;
/// the default `handle` method provides the shared forwarding logic
/// (this is the "template method" of the chain).
pub trait Approver {
    fn limit(&self) -> f64;
    fn name(&self) -> &str;
    fn next(&self) -> Option<&dyn Approver>;

    fn handle(&self, request: &PurchaseRequest) {
        if request.amount <= self.limit() {
            println!("  {} approved ${:.2} for {}", self.name(), request.amount, request.purpose);
        } else if let Some(next) = self.next() {
            next.handle(request);
        } else {
            println!(
                "  Request for ${:.2} ({}) could not be approved by anyone.",
                request.amount, request.purpose
            );
        }
    }
}
